test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java

```diff
@@ -0,0 +1,562 @@
+    }
+
+    /**
+     * Need to escape XML special characters.
+     */
+    private static void appendLine(StringBuilder builder, String line) {
+        if (line.contains("&")) {
+            line = line.replace("&amp;", "&");
+            line = line.replace("&lt;", "<");
+            line = line.replace("&gt;", ">");
+            line = line.replace("&quot;", "\"");
+            line = line.replace("&apos;", "'");
+        }
+        builder.append(line).append("\n");
+    }
+
+    private static int getCompileId(Matcher matcher) {
+        int compileId = -1;
+        try {
+            compileId = Integer.parseInt(matcher.group(1));
+        } catch (NumberFormatException e) {
+            TestRun.fail("Could not parse compile id", e);
+        }
+        return compileId;
+    }
+
+    /**
+     * Parse the compile id from this line if it belongs to a method that needs to be IR tested (part of test class
+     * and IR encoding from the test VM specifies that this method has @IR rules to be checked).
+     */
+    private void addTestMethodCompileId(Map<Integer, String> compileIdMap, String line) {
+        Pattern pattern = Pattern.compile("compile_id='(\\d+)'.*" + Pattern.quote(testClass.getCanonicalName()) + " (\\S+)");
+        Matcher matcher = pattern.matcher(line);
+        if (matcher.find()) {
+            // Only care about test class entries. Might have non-class entries as well if user specified additional
+            // compile commands. Ignore these.
+            String methodName = matcher.group(2);
+            if (compilations.containsKey(methodName)) {
+                // We only care about methods that we are actually gonna IR match based on IR encoding.
+                int compileId = getCompileId(matcher);
+                TestRun.check(!methodName.isEmpty(), "method name cannot be empty");
+                compileIdMap.put(compileId, methodName);
+            }
+        }
+    }
+
+    /**
+     * Make sure that line does not contain compile_kind which is used for OSR compilations which we are not
+     * interested in.
+     */
+    private static boolean isPrintIdealStart(String line) {
+        return line.startsWith("<ideal") && !line.contains("compile_kind='");
+    }
+
+    /**
+     * Make sure that line does not contain compile_kind which is used for OSR compilations which we are not
+     * interested in.
+     */
+    private static boolean isPrintOptoAssemblyStart(String line) {
+        return line.startsWith("<opto_assembly") && !line.contains("compile_kind='");
+    }
+
+    /**
+     * Get method name for this line by looking up the compile id.
+     * Returns null if not an interesting method (i.e. from test class).
+     */
+    private String getMethodName(Map<Integer, String> compileIdMap, String line) {
+        Pattern pattern = Pattern.compile("compile_id='(\\d+)'");
+        Matcher matcher = pattern.matcher(line);
```

Comment:
again, precompiled pattern can be saved into a (in this case) static field and reused.